File: strip-demo.js
#!/usr/bin/env node
/**
 * strip-demo.js — Promote a Relaunch demo folder into a paying-customer
 * production folder. Node.js port of strip-demo.py.
 *
 * Usage:
 *   node strip-demo.js <slug> --pin <6-digit-pin> --owner-email <email>
 *                              --allowed-origins URL [URL ...]
 *   node strip-demo.js <slug> --pin <new-pin> --rotate-only
 *   node strip-demo.js <slug> --pin <pin> --dry-run
 *
 * Examples:
 *   node strip-demo.js spork --pin 845984 --owner-email test@sporkbend.com \
 *     --allowed-origins https://client-spork.netlify.app
 *
 *   node strip-demo.js spork --pin 472913 --rotate-only
 */

const fs = require('fs');
const path = require('path');
const bcrypt = require('bcryptjs');

const REPO_ROOT = path.resolve(__dirname);
const CLIENTS_DIR = path.join(REPO_ROOT, 'clients');
const CONFIG_DIR = path.join(REPO_ROOT, 'config', 'clients');
const TEMPLATE_PATH = path.join(REPO_ROOT, 'templates', 'edit-page.html');

// ─────────────────────────────────────────────
// Bootstrap script template injected into each client's index.html.
// Reads JWT from localStorage; if valid + slug matches, sets up
// window.__relaunchWidgetConfig and loads tryrelaunch.com/widget.js.
// ─────────────────────────────────────────────
const WIDGET_BOOTSTRAP = `
<!-- Relaunch production edit widget bootstrap (auto-generated) -->
<script>
(function () {
  var SLUG = "__SLUG__";
  var FN_BASE = "https://tryrelaunch.com/.netlify/functions";
  var WIDGET_URL = "https://tryrelaunch.com/widget.js";
  var token;
  try { token = localStorage.getItem("relaunch_jwt"); } catch (e) { return; }
  if (!token) return;

  var payload;
  try {
    var parts = token.split(".");
    if (parts.length !== 3) return;
    var b64 = parts[1].replace(/-/g, "+").replace(/_/g, "/");
    while (b64.length % 4) b64 += "=";
    payload = JSON.parse(atob(b64));
  } catch (e) { return; }

  if (payload.exp && payload.exp * 1000 < Date.now()) {
    try { localStorage.removeItem("relaunch_jwt"); } catch (e) {}
    return;
  }
  if (payload.slug !== SLUG) return;

  window.__relaunchWidgetConfig = {
    slug: SLUG,
    token: token,
    endpoint: FN_BASE + "/edit-persistent",
    editPageUrl: "/edit/",
    supportEmail: "support@tryrelaunch.com"
  };

  var s = document.createElement("script");
  s.src = WIDGET_URL;
  s.async = true;
  document.head.appendChild(s);
})();
</script>
`;

// ─────────────────────────────────────────────
// CLI parsing (no external deps)
// ─────────────────────────────────────────────
function parseArgs(argv) {
  const args = {
    _: [],
    pin: null,
    ownerEmail: null,
    allowedOrigins: null,
    rotateOnly: false,
    dryRun: false
  };
  let i = 2;
  while (i < argv.length) {
    const arg = argv[i];
    if (arg === '--pin') { args.pin = argv[++i]; }
    else if (arg === '--owner-email') { args.ownerEmail = argv[++i]; }
    else if (arg === '--allowed-origins') {
      args.allowedOrigins = [];
      while (i + 1 < argv.length && !argv[i + 1].startsWith('--')) {
        args.allowedOrigins.push(argv[++i]);
      }
    }
    else if (arg === '--rotate-only') { args.rotateOnly = true; }
    else if (arg === '--dry-run') { args.dryRun = true; }
    else if (!arg.startsWith('--')) { args._.push(arg); }
    i++;
  }
  return args;
}

// ─────────────────────────────────────────────
// Strip helpers
// ─────────────────────────────────────────────
function stripBlockByOpeningTag(html, openingPattern, tag = 'div') {
  let count = 0;
  const openTok = `<${tag}`;
  const closeTok = `</${tag}>`;
  const validNext = new Set([' ', '>', '/', '\n', '\t', '\r']);

  while (true) {
    const m = html.match(openingPattern);
    if (!m) return [html, count];
    const start = m.index;
    let i = start + m[0].length;
    let depth = 1;
    while (i < html.length && depth > 0) {
      const nextOpen = html.indexOf(openTok, i);
      const nextClose = html.indexOf(closeTok, i);
      if (nextClose === -1) return [html, count];
      if (nextOpen !== -1 && nextOpen < nextClose) {
        const ch = html[nextOpen + openTok.length] || '';
        if (validNext.has(ch)) depth++;
        i = nextOpen + openTok.length;
      } else {
        depth--;
        i = nextClose + closeTok.length;
      }
    }
    if (depth !== 0) return [html, count];
    while (i < html.length && /\s/.test(html[i])) i++;
    html = html.slice(0, start) + html.slice(i);
    count++;
  }
}

function stripSimpleTag(html, openingPatternSrc, closing) {
  const escapedClosing = closing.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const fullPat = new RegExp(openingPatternSrc + '[\\s\\S]*?' + escapedClosing, 'g');
  let count = 0;
  const out = html.replace(fullPat, () => { count++; return ''; });
  return [out, count];
}

function stripCssSection(html, sectionName) {
  const escaped = sectionName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const openPat = new RegExp(
    '/\\*\\s*[\\u2500\\-=\\s]*\\s*' + escaped + '\\s*[\\u2500\\-=\\s]*\\*/',
    'i'
  );
  const m = html.match(openPat);
  if (!m) return [html, 0];
  const start = m.index;
  const after = m.index + m[0].length;
  const remainder = html.slice(after);
  const nextSection = remainder.match(/\/\*\s*[\u2500\-=]{2,}/);
  const styleEnd = remainder.indexOf('</style>');
  let end;
  if (nextSection) end = after + nextSection.index;
  else if (styleEnd !== -1) end = after + styleEnd;
  else return [html, 0];
  while (end > start && /\s/.test(html[end - 1])) end--;
  let out = html.slice(0, start) + html.slice(end);
  out = out.replace(/\n{3,}/g, '\n\n');
  return [out, 1];
}

// ─────────────────────────────────────────────
// Main HTML transform
// ─────────────────────────────────────────────
function transformHtml(html, slug, log) {
  let n;

  // 1. Remove robots noindex
  let result = html.replace(/\s*<meta\s+name="robots"\s+content="noindex"\s*\/?>\s*/g, '\n');
  if (result !== html) log.push('removed: noindex meta tag');
  html = result;

  // 2. SEO banner
  [html, n] = stripBlockByOpeningTag(html, /<div class="seo-banner">/);
  if (n) log.push(`removed: SEO banner div x${n}`);
  result = html.replace(/<!--\s*\u2554[\u2550\u2557]*[\s\S]*?SEO BANNER[\s\S]*?[\u2554\u2550\u255d\u255a\s]*-->/gi, () => '');
  if (result !== html) log.push('removed: SEO banner comment header');
  html = result;

  // 3. Sticky bar
  [html, n] = stripBlockByOpeningTag(html, /<div\s+class="sticky-bar(?:\s+show)?"\b/);
  if (n) log.push(`removed: sticky-bar div x${n}`);
  [html, n] = stripBlockByOpeningTag(html, /<div\s+(?:id="sticky-bar"|class="[^"]*sticky-bar[^"]*")\b/);
  if (n) log.push(`removed: sticky-bar (alt) x${n}`);

  // 4. Edit FAB
  [html, n] = stripSimpleTag(html, '<button\\s+class="edit-fab"[^>]*>', '</button>');
  if (n) log.push(`removed: button.edit-fab x${n}`);
  [html, n] = stripSimpleTag(html, '<button\\s+id="edit-fab"[^>]*>', '</button>');
  if (n) log.push(`removed: button#edit-fab x${n}`);

  // 5. Edit panel
  [html, n] = stripBlockByOpeningTag(html, /<div\s+class="edit-panel(?:\s+open)?"\b/);
  if (n) log.push(`removed: div.edit-panel x${n}`);
  [html, n] = stripBlockByOpeningTag(html, /<div\s+id="edit-panel"\b/);
  if (n) log.push(`removed: div#edit-panel x${n}`);

  // 6. CSS sections
  for (const section of ['SEO BANNER', 'STICKY CLAIM BAR', 'EDIT WIDGET']) {
    [html, n] = stripCssSection(html, section);
    if (n) log.push(`removed: CSS section [${section}]`);
  }

  // 7. Media-query rules for stripped elements
  result = html.replace(/\s*\.sticky-bar\s*\{[^}]*\}/g, '');
  if (result !== html) log.push('removed: .sticky-bar media-query rules');
  html = result;
  result = html.replace(/\s*\.edit-panel\s*\{[^}]*\}/g, '');
  if (result !== html) log.push('removed: .edit-panel media-query rules');
  html = result;

  // 8. Demo <script> blocks
  const demoTokens = [
    'toggleSEO', 'sendEdit', 'toggleEditPanel', 'closeBar',
    'edit-input', 'edit-messages', 'edit-typing', 'sticky-bar', 'editFab'
  ];
  let removedScripts = 0;
  result = html.replace(/<script\b[^>]*>[\s\S]*?<\/script>/g, (match) => {
    if (demoTokens.some(t => match.includes(t))) {
      removedScripts++;
      return '';
    }
    return match;
  });
  if (removedScripts) log.push(`removed: demo <script> blocks x${removedScripts}`);
  html = result;

  // 9. Strip previously-injected bootstrap (idempotent re-run)
  result = html.replace(/<!-- Relaunch production edit widget bootstrap[\s\S]*?<\/script>\s*/g, '');
  if (result !== html) log.push('removed: previous bootstrap (idempotent)');
  html = result;

  // 10. Inject production widget bootstrap before </body>
  const bootstrap = WIDGET_BOOTSTRAP.replace(/__SLUG__/g, slug);
  if (html.includes('</body>')) {
    html = html.replace('</body>', bootstrap + '\n</body>');
    log.push('injected: production widget bootstrap');
  } else {
    html = html + '\n' + bootstrap;
    log.push('appended: production widget bootstrap (no </body> found)');
  }

  // 11. Final cleanup
  html = html.replace(/\n{3,}/g, '\n\n');
  return html;
}

// ─────────────────────────────────────────────
// Edit page generation
// ─────────────────────────────────────────────
function generateEditPage(slug, log) {
  if (!fs.existsSync(TEMPLATE_PATH)) {
    throw new Error(`Edit page template not found at ${TEMPLATE_PATH}. Make sure templates/edit-page.html exists.`);
  }
  const template = fs.readFileSync(TEMPLATE_PATH, 'utf-8');
  const rendered = template.replace(/__SLUG__/g, slug);
  const editDir = path.join(CLIENTS_DIR, slug, 'edit');
  fs.mkdirSync(editDir, { recursive: true });
  fs.writeFileSync(path.join(editDir, 'index.html'), rendered, 'utf-8');
  log.push(`wrote: clients/${slug}/edit/index.html`);
}

// ─────────────────────────────────────────────
// Config write/update
// ─────────────────────────────────────────────
function writeConfig(slug, pin, ownerEmail, allowedOrigins, rotateOnly, log) {
  fs.mkdirSync(CONFIG_DIR, { recursive: true });
  const configPath = path.join(CONFIG_DIR, `${slug}.json`);
  const pinHash = bcrypt.hashSync(pin, 10);
  const nowIso = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');

  if (fs.existsSync(configPath) && rotateOnly) {
    const existing = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
    existing.pin_hash = pinHash;
    existing.pin_rotated_at = nowIso;
    fs.writeFileSync(configPath, JSON.stringify(existing, null, 2), 'utf-8');
    log.push(`updated: config/clients/${slug}.json (pin rotated)`);
    return;
  }

  let config;
  if (fs.existsSync(configPath)) {
    const existing = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
    existing.pin_hash = pinHash;
    existing.pin_rotated_at = nowIso;
    if (ownerEmail) existing.owner_email = ownerEmail;
    if (allowedOrigins) existing.allowed_origins = allowedOrigins;
    config = existing;
    log.push(`updated: config/clients/${slug}.json (preserving fields)`);
  } else {
    config = {
      slug,
      pin_hash: pinHash,
      owner_email: ownerEmail || null,
      allowed_origins: allowedOrigins || [],
      created_at: nowIso,
      pin_rotated_at: nowIso
    };
    log.push(`wrote: config/clients/${slug}.json`);
  }
  fs.writeFileSync(configPath, JSON.stringify(config, null, 2), 'utf-8');
}

// ─────────────────────────────────────────────
// Main
// ─────────────────────────────────────────────
function main() {
  const args = parseArgs(process.argv);
  const slug = (args._[0] || '').trim().toLowerCase();

  if (!slug || !/^[a-z0-9_-]+$/.test(slug)) {
    console.error("Usage:");
    console.error("  node strip-demo.js <slug> --pin <6-digit-pin> [--owner-email <email>] [--allowed-origins URL ...]");
    console.error("  node strip-demo.js <slug> --pin <new-pin> --rotate-only");
    console.error("  node strip-demo.js <slug> --pin <pin> --dry-run");
    process.exit(1);
  }
  if (!args.pin || !/^\d{6}$/.test(args.pin)) {
    console.error("ERROR: --pin must be exactly 6 digits.");
    process.exit(1);
  }

  const clientDir = path.join(CLIENTS_DIR, slug);
  if (!fs.existsSync(clientDir)) {
    console.error(`ERROR: clients/${slug}/ does not exist. Run xcopy /E /I previews\\${slug} clients\\${slug} first.`);
    process.exit(1);
  }

  const indexPath = path.join(clientDir, 'index.html');
  if (!fs.existsSync(indexPath) && !args.rotateOnly) {
    console.error(`ERROR: clients/${slug}/index.html does not exist.`);
    process.exit(1);
  }

  const log = [];

  if (args.rotateOnly) {
    if (args.dryRun) {
      console.log(`[DRY RUN] Would rotate PIN for ${slug}`);
      return;
    }
    writeConfig(slug, args.pin, args.ownerEmail, args.allowedOrigins, true, log);
    console.log(`\n✓ PIN rotated for ${slug}`);
    log.forEach(line => console.log(`  ${line}`));
    return;
  }

  const original = fs.readFileSync(indexPath, 'utf-8');
  const transformed = transformHtml(original, slug, log);

  if (args.dryRun) {
    console.log(`[DRY RUN] Original size: ${original.length.toLocaleString()} bytes`);
    console.log(`[DRY RUN] New size:      ${transformed.length.toLocaleString()} bytes`);
    log.forEach(line => console.log(`  ${line}`));
    console.log("[DRY RUN] Would also generate edit page and config.");
    return;
  }

  fs.writeFileSync(indexPath, transformed, 'utf-8');
  log.push(`wrote: clients/${slug}/index.html`);

  generateEditPage(slug, log);
  writeConfig(slug, args.pin, args.ownerEmail, args.allowedOrigins, false, log);

  console.log(`\n✓ Promoted ${slug} to production.`);
  log.forEach(line => console.log(`  ${line}`));
  console.log(`\nNext steps:`);
  console.log(`  1. Verify: open clients/${slug}/index.html in a browser to check the strip looks right`);
  console.log(`  2. Commit:`);
  console.log(`       git add clients/${slug}/ config/clients/${slug}.json`);
  console.log(`       git commit -m "Promote ${slug} to live client with PIN auth"`);
  console.log(`       git push origin main`);
  console.log(`  3. Test:`);
  console.log(`       Visit https://client-${slug}.netlify.app/edit/`);
  console.log(`       Enter PIN: ${args.pin}`);
}

main();

File: strip-demo.py
#!/usr/bin/env python3
"""
strip-demo.py v2 — Promote a Relaunch demo folder into a paying-customer
production folder.

Given a customer slug, this script:
  1. Strips demo-only elements from clients/<slug>/index.html
       (SEO banner, sticky CTA, demo widget, demo <script>)
  2. Injects the production edit-widget bootstrap loader (gates by JWT)
  3. Generates clients/<slug>/edit/index.html from the template
  4. Writes config/clients/<slug>.json with a bcrypt-hashed PIN

Usage:
  python strip-demo.py <slug> --pin <6-digit-pin>
                              --owner-email <email>
                              [--allowed-origins URL [URL ...]]
                              [--rotate-only]      # only update PIN; skip stripping/injecting
                              [--dry-run]          # print what would change

Examples:
  python strip-demo.py spork --pin 845984 --owner-email owner@sporkbend.com \\
      --allowed-origins https://client-spork.netlify.app https://sporkbend.com https://www.sporkbend.com

  python strip-demo.py spork --pin 472913 --rotate-only   # rotate PIN only

Requires:
  pip install bcrypt
"""
import argparse
import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path

try:
    import bcrypt
except ImportError:
    print("ERROR: bcrypt is required. Install with: pip install bcrypt")
    sys.exit(1)


REPO_ROOT = Path(__file__).resolve().parent
CLIENTS_DIR = REPO_ROOT / "clients"
CONFIG_DIR = REPO_ROOT / "config" / "clients"
TEMPLATE_PATH = REPO_ROOT / "templates" / "edit-page.html"


# ─────────────────────────────────────────────
# Bootstrap script that goes into each client's index.html.
# Reads JWT from localStorage; if valid + slug matches, sets up
# window.__relaunchWidgetConfig and loads tryrelaunch.com/widget.js.
# ─────────────────────────────────────────────
WIDGET_BOOTSTRAP_TEMPLATE = """\
<!-- Relaunch production edit widget bootstrap (auto-generated) -->
<script>
(function () {
  var SLUG = "__SLUG__";
  var FN_BASE = "https://tryrelaunch.com/.netlify/functions";
  var WIDGET_URL = "https://tryrelaunch.com/widget.js";
  var token;
  try { token = localStorage.getItem("relaunch_jwt"); } catch (e) { return; }
  if (!token) return;

  // Decode JWT payload (no signature check — server validates).
  var payload;
  try {
    var parts = token.split(".");
    if (parts.length !== 3) return;
    var b64 = parts[1].replace(/-/g, "+").replace(/_/g, "/");
    while (b64.length % 4) b64 += "=";
    payload = JSON.parse(atob(b64));
  } catch (e) { return; }

  if (payload.exp && payload.exp * 1000 < Date.now()) {
    try { localStorage.removeItem("relaunch_jwt"); } catch (e) {}
    return;
  }
  if (payload.slug !== SLUG) return;

  window.__relaunchWidgetConfig = {
    slug: SLUG,
    token: token,
    endpoint: FN_BASE + "/edit-persistent",
    editPageUrl: "/edit/",
    supportEmail: "support@tryrelaunch.com"
  };

  var s = document.createElement("script");
  s.src = WIDGET_URL;
  s.async = true;
  document.head.appendChild(s);
})();
</script>
"""


# ─────────────────────────────────────────────
# Helpers — same approach as v1 but tightened
# ─────────────────────────────────────────────

def strip_block_by_opening_tag(html, opening_pattern, tag='div'):
    """
    Find an opening tag matching opening_pattern, then walk forward
    counting <tag> opens and </tag> closes until depth returns to zero.
    Removes the entire block. Returns (new_html, count_removed).
    """
    count = 0
    while True:
        m = re.search(opening_pattern, html)
        if not m:
            return html, count
        start = m.start()
        pos = m.end()
        depth = 1
        i = pos
        open_tok = f'<{tag}'
        close_tok = f'</{tag}>'
        while i < len(html) and depth > 0:
            next_open = html.find(open_tok, i)
            next_close = html.find(close_tok, i)
            if next_close == -1:
                return html, count  # malformed; bail
            if next_open != -1 and next_open < next_close:
                ch = html[next_open + len(open_tok):next_open + len(open_tok) + 1]
                if ch in (' ', '>', '/', '\n', '\t'):
                    depth += 1
                i = next_open + len(open_tok)
            else:
                depth -= 1
                i = next_close + len(close_tok)
        if depth != 0:
            return html, count
        while i < len(html) and html[i] in ' \t\r\n':
            i += 1
        html = html[:start] + html[i:]
        count += 1


def strip_simple_tag(html, opening_pattern, closing):
    pattern = opening_pattern + r'.*?' + re.escape(closing)
    new_html, n = re.subn(pattern, '', html, flags=re.DOTALL)
    return new_html, n


def strip_css_section(html, section_name):
    open_pat = re.compile(
        r'/\*\s*[─\-=\s]*\s*' + re.escape(section_name) + r'\s*[─\-=\s]*\*/',
        re.IGNORECASE
    )
    m = open_pat.search(html)
    if not m:
        return html, 0
    start = m.start()
    after = m.end()
    next_section = re.search(r'/\*\s*[─\-=]{2,}', html[after:])
    style_end = html.find('</style>', after)
    if next_section is not None:
        end = after + next_section.start()
    elif style_end != -1:
        end = style_end
    else:
        return html, 0
    while end > start and html[end - 1] in ' \t\r\n':
        end -= 1
    new_html = html[:start] + html[end:]
    new_html = re.sub(r'\n{3,}', '\n\n', new_html)
    return new_html, 1


# ─────────────────────────────────────────────
# Strip demo elements (same as v1) + inject production bootstrap
# ─────────────────────────────────────────────

def transform_html(html, slug, log):
    # 1. Remove robots noindex meta — production sites should be indexed
    html, n = re.subn(
        r'\s*<meta\s+name="robots"\s+content="noindex"\s*/?>\s*',
        '\n',
        html
    )
    if n: log.append(f"removed: noindex meta tag x{n}")

    # 2. SEO banner block + its comment header
    html, n = strip_block_by_opening_tag(html, r'<div class="seo-banner">')
    if n: log.append(f"removed: SEO banner div x{n}")
    html, n = re.subn(
        r'<!--\s*╔[═╗]*[^╝]*?SEO BANNER[^╗]*?[╔═╝╚\s]*-->',
        '', html, flags=re.IGNORECASE | re.DOTALL
    )
    if n: log.append(f"removed: SEO banner comment header x{n}")

    # 3. Sticky bar
    html, n = strip_block_by_opening_tag(html, r'<div\s+class="sticky-bar(?:\s+show)?"\b')
    if n: log.append(f"removed: sticky-bar div x{n}")
    html, n = strip_block_by_opening_tag(
        html, r'<div\s+(?:id="sticky-bar"|class="[^"]*sticky-bar[^"]*")\b'
    )
    if n: log.append(f"removed: sticky-bar (id/alt class) x{n}")

    # 4. Edit FAB
    html, n = strip_simple_tag(html, r'<button\s+class="edit-fab"[^>]*>', '</button>')
    if n: log.append(f"removed: <button.edit-fab> x{n}")
    html, n = strip_simple_tag(html, r'<button\s+id="edit-fab"[^>]*>', '</button>')
    if n: log.append(f"removed: <button#edit-fab> x{n}")

    # 5. Edit panel
    html, n = strip_block_by_opening_tag(html, r'<div\s+class="edit-panel(?:\s+open)?"\b')
    if n: log.append(f"removed: <div.edit-panel> x{n}")
    html, n = strip_block_by_opening_tag(html, r'<div\s+id="edit-panel"\b')
    if n: log.append(f"removed: <div#edit-panel> x{n}")

    # 6. CSS sections
    for section in ['SEO BANNER', 'STICKY CLAIM BAR', 'EDIT WIDGET']:
        html, n = strip_css_section(html, section)
        if n: log.append(f"removed: CSS section [{section}]")

    # 7. Media-query rules for stripped elements
    html, n = re.subn(r'\s*\.sticky-bar\s*\{[^}]*\}', '', html)
    if n: log.append(f"removed: .sticky-bar media-query rules x{n}")
    html, n = re.subn(r'\s*\.edit-panel\s*\{[^}]*\}', '', html)
    if n: log.append(f"removed: .edit-panel media-query rules x{n}")

    # 8. Demo <script> blocks (anything that mentions demo function names)
    def is_demo_script(match):
        body = match.group(0)
        return any(token in body for token in (
            'toggleSEO', 'sendEdit', 'toggleEditPanel',
            'closeBar', 'edit-input', 'edit-messages',
            'edit-typing', 'sticky-bar', 'editFab'
        ))

    script_pat = re.compile(r'<script\b[^>]*>.*?</script>', re.DOTALL)
    new_parts = []
    last_end = 0
    removed_scripts = 0
    for m in script_pat.finditer(html):
        if is_demo_script(m):
            new_parts.append(html[last_end:m.start()])
            last_end = m.end()
            removed_scripts += 1
    new_parts.append(html[last_end:])
    if removed_scripts:
        html = ''.join(new_parts)
        log.append(f"removed: demo <script> blocks x{removed_scripts}")

    # 9. Strip any previously-injected bootstrap (idempotent re-run)
    html, n = re.subn(
        r'<!-- Relaunch production edit widget bootstrap.*?</script>\s*',
        '', html, flags=re.DOTALL
    )
    if n: log.append(f"removed: previous bootstrap (idempotent) x{n}")

    # 10. Inject production widget bootstrap before </body>
    bootstrap = WIDGET_BOOTSTRAP_TEMPLATE.replace('__SLUG__', slug)
    if '</body>' in html:
        html = html.replace('</body>', bootstrap + '\n</body>', 1)
        log.append("injected: production widget bootstrap")
    else:
        html = html + '\n' + bootstrap
        log.append("appended: production widget bootstrap (no </body> found)")

    # 11. Final cleanup
    html = re.sub(r'\n{3,}', '\n\n', html)
    return html


# ─────────────────────────────────────────────
# Generate edit/index.html from template
# ─────────────────────────────────────────────

def generate_edit_page(slug, log):
    if not TEMPLATE_PATH.exists():
        raise FileNotFoundError(
            f"Edit page template not found at {TEMPLATE_PATH}. "
            "Make sure templates/edit-page.html exists."
        )
    template = TEMPLATE_PATH.read_text(encoding='utf-8')
    rendered = template.replace('__SLUG__', slug)
    edit_dir = CLIENTS_DIR / slug / 'edit'
    edit_dir.mkdir(parents=True, exist_ok=True)
    edit_path = edit_dir / 'index.html'
    edit_path.write_text(rendered, encoding='utf-8')
    log.append(f"wrote: clients/{slug}/edit/index.html")


# ─────────────────────────────────────────────
# Write/update client config
# ─────────────────────────────────────────────

def write_config(slug, pin, owner_email, allowed_origins, log, rotate_only=False):
    CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    config_path = CONFIG_DIR / f"{slug}.json"

    pin_hash = bcrypt.hashpw(pin.encode('utf-8'), bcrypt.gensalt(rounds=10)).decode('utf-8')
    now_iso = datetime.now(timezone.utc).isoformat(timespec='seconds').replace('+00:00', 'Z')

    if config_path.exists() and rotate_only:
        existing = json.loads(config_path.read_text(encoding='utf-8'))
        existing['pin_hash'] = pin_hash
        existing['pin_rotated_at'] = now_iso
        config_path.write_text(json.dumps(existing, indent=2), encoding='utf-8')
        log.append(f"updated: config/clients/{slug}.json (pin rotated)")
        return existing

    new_config = {
        "slug": slug,
        "pin_hash": pin_hash,
        "owner_email": owner_email or None,
        "allowed_origins": allowed_origins or [],
        "created_at": now_iso,
        "pin_rotated_at": now_iso
    }

    if config_path.exists():
        existing = json.loads(config_path.read_text(encoding='utf-8'))
        existing['pin_hash'] = pin_hash
        existing['pin_rotated_at'] = now_iso
        if owner_email:
            existing['owner_email'] = owner_email
        if allowed_origins:
            existing['allowed_origins'] = allowed_origins
        config_path.write_text(json.dumps(existing, indent=2), encoding='utf-8')
        log.append(f"updated: config/clients/{slug}.json (preserving existing fields)")
        return existing

    config_path.write_text(json.dumps(new_config, indent=2), encoding='utf-8')
    log.append(f"wrote: config/clients/{slug}.json")
    return new_config


# ─────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Promote a demo to a paying client.")
    parser.add_argument('slug', help='Client slug (e.g., spork)')
    parser.add_argument('--pin', required=True, help='6-digit numeric PIN')
    parser.add_argument('--owner-email', default=None, help='Customer email')
    parser.add_argument('--allowed-origins', nargs='+', default=None,
                        help='List of full origin URLs allowed for CORS '
                             '(e.g. https://client-spork.netlify.app https://sporkbend.com)')
    parser.add_argument('--rotate-only', action='store_true',
                        help='Only update the PIN — skip stripping/injecting HTML')
    parser.add_argument('--dry-run', action='store_true',
                        help="Don't write files; just print what would happen")
    args = parser.parse_args()

    slug = args.slug.strip().lower()
    if not re.match(r'^[a-z0-9_-]+$', slug):
        print(f"ERROR: invalid slug '{slug}'. Use only a-z, 0-9, dash, underscore.")
        sys.exit(1)

    if not re.match(r'^\d{6}$', args.pin):
        print(f"ERROR: --pin must be exactly 6 digits.")
        sys.exit(1)

    client_dir = CLIENTS_DIR / slug
    if not client_dir.exists():
        print(f"ERROR: clients/{slug}/ does not exist. Run xcopy /E /I previews\\{slug} clients\\{slug} first.")
        sys.exit(1)

    index_path = client_dir / 'index.html'
    if not index_path.exists() and not args.rotate_only:
        print(f"ERROR: clients/{slug}/index.html does not exist.")
        sys.exit(1)

    log = []

    if args.rotate_only:
        if args.dry_run:
            print(f"[DRY RUN] Would rotate PIN for {slug}")
            return
        write_config(slug, args.pin, args.owner_email, args.allowed_origins,
                     log, rotate_only=True)
        print(f"\n✓ PIN rotated for {slug}")
        for line in log:
            print(f"  {line}")
        return

    # Full promotion
    original = index_path.read_text(encoding='utf-8')
    transformed = transform_html(original, slug, log)

    if args.dry_run:
        print(f"[DRY RUN] Would transform clients/{slug}/index.html")
        print(f"[DRY RUN] Original size: {len(original):,} bytes")
        print(f"[DRY RUN] New size:      {len(transformed):,} bytes")
        for line in log:
            print(f"  {line}")
        print("[DRY RUN] Would also generate edit page and config.")
        return

    index_path.write_text(transformed, encoding='utf-8')
    log.append(f"wrote: clients/{slug}/index.html")

    generate_edit_page(slug, log)
    write_config(slug, args.pin, args.owner_email, args.allowed_origins, log)

    print(f"\n✓ Promoted {slug} to production.")
    for line in log:
        print(f"  {line}")
    print(f"\nNext steps:")
    print(f"  1. Review: open clients/{slug}/index.html in a browser to verify the strip looks right")
    print(f"  2. Commit:")
    print(f"       git add clients/{slug}/ config/clients/{slug}.json")
    print(f'       git commit -m "Promote {slug} to live client with PIN auth"')
    print(f"       git push origin main")
    print(f"  3. Test:")
    print(f"       Visit https://client-{slug}.netlify.app/edit/")
    print(f"       Enter PIN: {args.pin}")
    print(f"       Should redirect to homepage with edit widget appearing.")


if __name__ == '__main__':
    main()
